### PR TITLE
c#, improvement: error parsing 

### PR DESCRIPTION
**Pull request by dcb6**
We now generate Exception types for all errors that are defined in the IR. Generated clients with an error discrimination strategy of "status code" will throw one of these typed Exceptions based on the status code of error responses. Example error type:
  ```csharp
  public sealed class UnauthorizedRequest(UnauthorizedRequestErrorBody body)
      : MyCompanyApiException("UnauthorizedRequest", 401, body)
  {
      public new UnauthorizedRequestErrorBody Body { get; } = body;
  }
  ```

### CODE REVIEW

File: generators/csharp/sdk/src/SdkGeneratorContext.ts [hunk 1]
```diff
@@ -111,6 +122,31 @@ export class SdkGeneratorContext extends AbstractCsharpGeneratorContext<SdkCusto
...
+        return csharp.classReference({
+            name: this.getExceptionPrefix() + "ApiException",
+            namespace: this.getCoreNamespace()
+        });
+    }
+
+    public getExceptionClassReference(declaredErrorName: DeclaredErrorName): csharp.ClassReference {
+        return csharp.classReference({
+            name: this.getPascalCaseSafeName(declaredErrorName.name),
+            namespace: this.getNamespaceFromFernFilepath(declaredErrorName.fernFilepath)
```
Comment (dcb6) [inline generators/csharp/sdk/src/SdkGeneratorContext.ts:142]: @dsinghvi i know we very briefly discussed if we should have the `explicit-namespaces` config apply here and landed on `no`

fysa i've reversed on that and think it's easiest and most consistent (and not too consequential) to apply the config

File: seed/csharp-sdk/alias/src/SeedAlias/SeedAliasClient.cs [hunk 2]
```diff
@@ -21,13 +21,23 @@ public SeedAliasClient(ClientOptions? clientOptions = null)
...
-        await _client.MakeRequestAsync(
+        var response = await _client.MakeRequestAsync(
             new RawClient.JsonApiRequest
             {
                 BaseUrl = _client.Options.BaseUrl,
                 Method = HttpMethod.Get,
                 Path = $"/{typeId}"
             }
         );
+        if (response.StatusCode is >= 200 and < 400)
```
Comment (dcb6) [inline seed/csharp-sdk/alias/src/SeedAlias/SeedAliasClient.cs:32]: note this change -- we previously weren't even checking response status for endpoints without a response 😲

File: seed/csharp-sdk/custom-auth/src/SeedCustomAuth/CustomAuth/CustomAuthClient.cs [hunk 3]
```diff
@@ -52,8 +80,36 @@ public async Task<bool> PostWithCustomAuthAsync(object request)
...
+            }
+            catch (JsonException e)
+            {
+                throw new SeedCustomAuthException("Failed to deserialize response", e);
+            }
+        }
+
+        try
+        {
+            switch (response.StatusCode)
```
Comment (dcb6) [inline seed/csharp-sdk/custom-auth/src/SeedCustomAuth/CustomAuth/CustomAuthClient.cs:95]: this is a decent one to check out!

File: generators/csharp/sdk/src/error/BaseExceptionGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { csharp, CSharpFile, FileGenerator } from "@fern-api/csharp-codegen";
+import { join, RelativeFilePath } from "@fern-api/fs-utils";
+import { SdkCustomConfigSchema } from "../SdkCustomConfig";
+import { SdkGeneratorContext } from "../SdkGeneratorContext";
+
+export class BaseExceptionGenerator extends FileGenerator<CSharpFile, SdkCustomConfigSchema, SdkGeneratorContext> {
+    constructor(context: SdkGeneratorContext) {
```
Comment (amckinney) [inline generators/csharp/sdk/src/error/BaseExceptionGenerator.ts:7]: Looks like we can drop this according to the lint warning below.

**Review by amckinney — COMMENTED** [review]
Looks great overall, mostly nits and only a couple things to consider (e.g. the `= null` check) before stamp.

File: generators/csharp/codegen/src/ast/Class.ts [hunk 5]
```diff
@@ -138,6 +166,17 @@ export class Class extends AstNode {
...
         writer.write(this.record ? " record" : " class");
         writer.write(` ${this.name}`);
+        if (this.primaryConstructor != null && this.primaryConstructor.parameters.length > 0) {
+            const primaryConstructor = this.primaryConstructor;
+            writer.write("(");
+            primaryConstructor.parameters.forEach((parameter, index) => {
+                parameter.write(writer);
+                if (index < primaryConstructor.parameters.length - 1) {
+                    writer.write(", ");
+                }
```
[the remark this answers is not in the document]
Comment (dcb6) [reply generators/csharp/codegen/src/ast/Class.ts:176]: 🧠 amazing

File: generators/csharp/sdk/src/endpoint/EndpointGenerator.ts [hunk 6]
```diff
@@ -167,41 +166,120 @@ export class EndpointGenerator {
...
+                    writer.writeNode(this.context.getExceptionClassReference(fullError.name));
+                    writer.write("(");
+                    writer.writeNode(this.context.getJsonUtilsClassReference());
+                    writer.write(".Deserialize<");
+                    writer.writeNode(
+                        fullError.type != null
+                            ? this.context.csharpTypeMapper.convert({ reference: fullError.type })
+                            : csharp.Type.object()
+                    );
+                    writer.writeTextStatement(`>(${RESPONSE_BODY_VARIABLE_NAME}))`);
```
[the remark this answers is not in the document]
Comment (dcb6) [reply generators/csharp/sdk/src/endpoint/EndpointGenerator.ts:202]: are you imagining it being re-used, or just trying to modularize a bit?

File: generators/csharp/sdk/src/error/BaseExceptionGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            parentClassReference: csharp.classReference({ name: "Exception", namespace: "System" }),
+            primaryConstructor: {
+                parameters: [
+                    csharp.parameter({ name: "message", type: csharp.Type.string() }),
+                    csharp.parameter({
+                        name: "innerException",
+                        type: csharp.Type.optional(
+                            csharp.Type.reference(csharp.classReference({ name: "Exception", namespace: "System" }))
+                        ),
+                        initializer: "null"
```
[the remark this answers is not in the document]
Comment (dcb6) [reply generators/csharp/sdk/src/error/BaseExceptionGenerator.ts:24]: discussed offline and found that it is needed!

File: generators/csharp/sdk/src/endpoint/EndpointGenerator.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (amckinney) [reply generators/csharp/sdk/src/endpoint/EndpointGenerator.ts:202]: Just trying to modularize - totally unnecessary if you don't think it's useful!

**Review by amckinney — APPROVED** [review]
Now that the `= null` bit is resolved, this lgtm once some of the nits are resolved!

File: generators/csharp/sdk/src/endpoint/EndpointGenerator.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcb6) [reply generators/csharp/sdk/src/endpoint/EndpointGenerator.ts]: i like it, done!

**Pull request merged**